rust/connlib/tunnel/src/client.rs

```diff
@@ -284,6 +272,38 @@ where
 
         Ok(())
     }
+
+    fn update_dns_mapping(&mut self) -> bool {
+        let Some(config) = &self.role_state.interface_config else {
+            return false;
+        };
+
+        let effective_dns_servers = effective_dns_servers(
+            config.upstream_dns.clone(),
+            self.role_state.system_resolvers.clone(),
+        );
```

Comment:
These are the ~~current~~ new effective DNS servers

This function is called after the DNS servers are updated in `role_state`? So the role state is out of sync until its owner calls this?

And this is where the filtering happens, in `effective_dns_servers`, and that also means we won't spuriously update anything if the system DNS changes while an upstream DNS is set.

Reply:
well, the system's resolvers are right before this in `role_state` but we haven't updated the dns mappings considering the newest one.

> And this is where the filtering happens, in effective_dns_servers, and that also means we won't spuriously update anything if the system DNS changes while an upstream DNS is set.

yup!